### PR TITLE
8335444: Generalize implementation of AndNode mul_ring

### PR DESCRIPTION
**Pull request by jaskarth**
Hi all,
I've written this patch which improves type calculation for bitwise-and functions. Previously, the only cases that were considered were if one of the inputs to the node were a positive constant. I've generalized this behavior, as well as added a case to better estimate the result for arbitrary ranges. Since these are very common patterns to see, this can help propagate more precise types throughout the ideal graph for a large number of methods, making other optimizations and analyses stronger. I was interested in where this patch improves types, so I ran CTW for `java_base` and `java_base_2` and printed out the differences in this gist [here](https://gist.github.com/jaskarth/b45260d81ab621656f4a55cc51cf5292). While I don't think it's particularly complicated I've also added some discussion of the mathematics below, mostly because I thought it was interesting to work through :)

This patch passes tier1-3 testing on my linux x64 machine. Thoughts and reviews would be very appreciated!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335444](https://bugs.openjdk.org/browse/JDK-8335444): Generalize implementation of AndNode mul_ring (**Enhancement** - P4)

### Reviewers
 * [Damon Fenacci](https://openjdk.org/census#dfenacci) (@dafedafe - Committer) Review applies to [ca2db583](https://git.openjdk.org/jdk/pull/20066/files/ca2db5839a1126232cb08cb3c09ee12bb32acb34)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20066/head:pull/20066` \
`$ git checkout pull/20066`

Update a local copy of the PR: \
`$ git checkout pull/20066` \
`$ git pull https://git.openjdk.org/jdk.git pull/20066/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20066`

View PR using the GUI difftool: \
`$ git pr show -t 20066`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20066.diff">https://git.openjdk.org/jdk/pull/20066.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20066#issuecomment-2212936412)

### CODE REVIEW

**Comment by jaskarth** [comment]
The common cases I saw in java.base and other real world programs that aren't handled currently are with positive ranges, negative constants, and ranges that cross 0. For positive ranges it's possible to generalize the existing case with a positive constant, and for negative constants and ranges that cross 0 I came up with an estimation.

For a positive constant $C$, we can say that for any integer type $I$ that the result of the operation $`I \mathbin{\&} C`$ will be bounded by the range $[0, C]$. This can then be generalized to positive ranges, as for any constant $c_{1}$ such that $0 \le c_{1} \le C$ the result of $`I \mathbin{\&} c_{1}`$ will also be bounded by $[0, C]$. Therefore, we can say that $`I \mathbin{\&} [c_{1}, C]`$ for any $0 \le c_{1} \le C$ will be in the range $[0, C]$. This case can be further generalized if we know both ranges are positive. For the operation $`[L_{1}, U_{1}] \mathbin{\&} [L_{2}, U_{2}]`$ where $L_{1} \ge 0$ and $L_{2} \ge 0$ the result will be bounded by $`[0, \text{min}(U_{1}, U_{2})]`$. Since bitwise-and can only remove bits, we know that the maximum possible value of the result will be determined by the range that has fewer bits set.

In the case of negative constants and ranges that cross 0 it's more difficult to numerically derive a solution due to two's complement, as the minimum value can be lower than both of the ranges' minimum values. Since a negative two's complement number will always begin with a sequence of 1s, and that sequence of 1s grows smaller as the number becomes lower, the lower bound will be the number that contains the sequence of ones of the smaller value of both ranges. For the upper bound, the result will be the higher value of both ranges, as the result of bitwise-and with a negative value can return a result up to the positive value.

While these methods produce much more precise types for many cases, they are by no means optimal. For that I think proper bitwise propagation is needed, such as with #17508. However, I think this is still worthwhile to get in since it's an easy win for most cases and is a pretty simple patch.

**Comment by chhagedorn** [comment]
Not a review but I quickly ran it through our testing and the following test fails with `-XX:UseAVX=3` on linux-x64-debug and windows-x64-debug and without any flags on windows-x64-debug which seems to be related to your patch:

Test: compiler/vectorization/runner/BasicBooleanOpTest.java

Output:
```
Failed IR Rules (1) of Methods (1)
----------------------------------
1) Method "public boolean[] compiler.vectorization.runner.BasicBooleanOpTest.vectorAnd()" - [Failed IR rules: 1]:
   * @IR rule 3: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#MACRO_LOGIC_V#_", ">0"}, failOn={}, applyIfPlatform={}, applyIfPlatformOr={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={"avx512f", "true"}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "MacroLogicV"
           - Failed comparison: [found] 0 > 0 [given]
           - No nodes matched!
```

**Review by dafedafe — CHANGES_REQUESTED** [review]

**Comment by jaskarth** [comment]
Thank you for running testing @chhagedorn! I think I didn't run into this because my device doesn't support AVX-512. Does the failure have an ideal node printout as well? I think that could help as well in diagnosing the issue. Thanks!

File: src/hotspot/share/opto/mulnode.cpp
Comment (dafedafe) [reply src/hotspot/share/opto/mulnode.cpp:638]: You're totally right! It is not even related to the LSB (`-14 & -6` would have the same problem with `-12`). Finding the leading 1s is the right solution. Thanks a lot for the clarification!

**Comment by chhagedorn** [comment]
> Thank you for running testing @chhagedorn! I think I didn't run into this because my device doesn't support AVX-512. Does the failure have an ideal node printout as well? I think that could help in diagnosing the issue. Thanks!

Sure, here is the log file for linux-x64-debug: [test_failure.log](https://github.com/user-attachments/files/16445886/test_failure.log)

**Comment by dafedafe** [comment]
@jaskarth out of curiosity: could you by chance notice any measurable performance difference (e.g. for specific/ad-hoc benchmarks)?

**Comment by jaskarth** [comment]
@chhagedorn I've had a chance to investigate the IR test failure, and it seems it's because with the patch we can find a sharper type during IGVN than with the baseline. After parsing, the code shape inside the loop looks like this: `StoreB[idx] = (a & b) & 1;` As the StoreB is storing a boolean type, we perform a bitwise-and of 1 to ensure the boolean value is in bounds. With the patch, the type of `(a & b)` is `bool` so the `& 1` is removed in `AndINode::Identity`. In the baseline, the type of `(a & b)` is `int`, so the redundant AndNode isn't removed. Later on, the logic chain is transformed into a `MacroLogicVNode`.

Since the IR output of the operation has changed I've updated the IR checks, bringing it in line with the other operations in the test.

**Comment by jaskarth** [comment]
@dafedafe I added a microbenchmark based on the case I saw above, and got these results:
```java
                                                                                             Baseline                    Patch           Improvement
Benchmark                                                   (COUNT)  (seed)  Mode  Cnt    Score   Error  Units     Score   Error  Units
TypeVectorOperations.TypeVectorOperationsNonSuperWord.andZ      512       0  avgt    8  155.288 ± 1.175  ns/op   188.844 ± 4.189  ns/op  (+ 19.5%)
TypeVectorOperations.TypeVectorOperationsNonSuperWord.andZ     2048       0  avgt    8  629.098 ± 7.489  ns/op   732.558 ± 3.983  ns/op  (+ 15.2%)
TypeVectorOperations.TypeVectorOperationsSuperWord.andZ         512       0  avgt    8   22.917 ± 0.338  ns/op    23.578 ± 1.003  ns/op  (+ 2.8%)
TypeVectorOperations.TypeVectorOperationsSuperWord.andZ        2048       0  avgt    8   35.683 ± 0.232  ns/op    37.546 ± 1.063  ns/op  (+ 5.1%)
```
In general though I've found that unfortunately it's pretty difficult to identify specific places where performance is improved, since rather than improving nodes locally this analysis strengthens other idealizations that use int types. By improving the type we might be able to find more operations that evaluate to constants or prune out redundant comparisons, either directly or through another node that transforms the type further. I've been wanting to make our type analysis stronger, so that we can find more nontrivial optimizations without needing specialized idealization rules.

**Comment by chhagedorn** [comment]
> @chhagedorn I've had a chance to investigate the IR test failure, and it seems it's because with the patch we can find a sharper type during IGVN than with the baseline. After parsing, the code shape inside the loop looks like this: `StoreB[idx] = (a & b) & 1;` As the StoreB is storing a boolean type, we perform a bitwise-and of 1 to ensure the boolean value is in bounds. With the patch, the type of `(a & b)` is `bool` so the `& 1` is removed in `AndINode::Identity`. In the baseline, the type of `(a & b)` is `int`, so the redundant AndNode isn't removed. Later on, the logic chain is transformed into a `MacroLogicVNode`.
> 
> Since the IR output of the operation has changed I've updated the IR checks, bringing it in line with the other operations in the test.

Thanks for the investigation and the update! That sounds reasonable. Let me submit some testing again for that test.

**Comment by chhagedorn** [comment]
The test is still failing, with `UseAVX=3` and without. It seems that `MacroLogicVNode` is present in both logs:
[normal.log](https://github.com/user-attachments/files/16507041/normal.log)
[avx3.log](https://github.com/user-attachments/files/16507059/avx3.log)

**Comment by jaskarth** [comment]
Hmm, that failure is quite peculiar, since now it seems we're failing because we're creating extra `MacroLogicV` nodes rather than failing because we're not creating any. Unfortunately, I didn't have much luck debugging the root cause since I don't have access to AVX-512 hardware. I've changed the IR check to a phase before `MacroLogicV` nodes are created, which should hopefully fix the failure.

**Comment by chhagedorn** [comment]
> Hmm, that failure is quite peculiar, since now it seems we're failing because we're creating extra `MacroLogicV` nodes rather than failing because we're not creating any. Unfortunately, I didn't have much luck debugging the root cause since I don't have access to AVX-512 hardware. I've changed the IR check to a phase before `MacroLogicV` nodes are created, which should hopefully fix the failure.

Even if you don't have access to real hardware, you could try to emulate it with [Intel SDE](https://www.intel.com/content/www/us/en/developer/articles/tool/software-development-emulator.html) to see what's going on. Might be worth a shot. Nevertheless, I will have a look at your patch again in closer detail tomorrow or hopefully on Monday.

**Comment by jaskarth** [comment]
> you could try to emulate it with [Intel SDE](https://www.intel.com/content/www/us/en/developer/articles/tool/software-development-emulator.html) to see what's going on

Oh I hadn't heard of that tool before, I'll give it a try. Thank you for the link!

File: src/hotspot/share/opto/mulnode.cpp
Comment (chhagedorn) [inline src/hotspot/share/opto/mulnode.cpp:635]: `sel_val` can only be 0 if `r0->_lo` and `r1->_lo` are both -1. While I think it's correct how you handle the case here, wouldn't it be simpler/more readable if you handle this case separately by setting -1 as lower bound directly instead of using "`min >> #digits"`?

**Review by chhagedorn — COMMENTED** [review]
Interesting findings! I like that you unified the integer and long case with a template. A few comments.

File: src/hotspot/share/opto/mulnode.cpp
Comment (merykitty) [inline src/hotspot/share/opto/mulnode.cpp:601]:
```suggestion
  int widen = MAX2(r0->_widen, r1->_widen);
```

**Review by merykitty — COMMENTED** [review]
This is really nice. Given Christian's suggestion, what do you think about inferring the unsigned bounds and known bits from the signed bounds, do the calculation there and work back the signed bounds in the end?

File: src/hotspot/share/opto/mulnode.cpp
Comment (merykitty) [reply src/hotspot/share/opto/mulnode.cpp:601]: For example, since `res_uhi = min(uhi1, uhi2)`, your cases regarding one of the inputs being non-negative can be nicely arrived at because for a non-negative `TypeInt`, its unsigned upper bound is the same as its signed upper bound.

File: src/hotspot/share/opto/mulnode.cpp
Comment (jaskarth) [reply src/hotspot/share/opto/mulnode.cpp:638]: This is a great observation! Since bitwise-and can only remove bits, the largest possible value is the smaller of each range's `hi` value so I think it's correct to use the minimum here rather than the maximum. I didn't look into this case too deeply initially since I didn't find any bitwise-and nodes of two negative ranges in my investigation, but I think we should include it since it's a simple enough condition to check.

File: src/hotspot/share/opto/mulnode.cpp
Comment (jaskarth) [reply src/hotspot/share/opto/mulnode.cpp:611]: I agree, I think it would be good to include the math in the code to make it easier to understand. I'll add it in the next commit.

File: src/hotspot/share/opto/mulnode.cpp
Comment (jaskarth) [reply src/hotspot/share/opto/mulnode.cpp:635]: This is a good point! The logic becomes a lot easier to parse when we don't need to work around `count_leading_zeros` being undefined at 0 and just set the lower bound manually.

**Comment by jaskarth** [comment]
Thanks for taking a look, @merykitty! With regard to unsigned and bitwise bounds I was thinking for this patch it might be better to keep everything in the signed domain to keep the patch simple, since AFAIK the analysis to go between the signed domain and the unsigned/bitwise domain can become quite complicated. I think it might be better to do that sort of analysis more generally in the type system, such as what you're doing with #17508. I was thinking this PR could be a sort of baseline for any future improvements using that system. Let me know what you think about this approach, or if you think I should explore unsigned and bitwise bounds in this patch. Thanks!

**Review by merykitty — APPROVED** [review]
I agree, I think your math is correct.

File: src/hotspot/share/opto/mulnode.cpp
Comment (merykitty) [reply src/hotspot/share/opto/mulnode.cpp:638]: Another point of view, you can decompose `[lo, hi]` into `[lo, -1] v [0, hi]`. Then `[lo1, hi1] & [lo2, hi2]` can be calculated as:

    ([lo1, -1] & [lo2, -1]) v ([lo1, -1] & [0, hi2]) v ([0, hi1] & [lo2, -1]) v ([0, hi1] & [0, hi2]) =
    [lo, -1] v [0, hi2] v [0, h1] v [0, min(hi1, hi2)] =
    [lo, max(hi1, hi2)]`

with `lo` being the lower bound you calculated right above.

Comment (merykitty) [reply src/hotspot/share/opto/mulnode.cpp:638]: Just to be clear I think `MAX2` is the correct thing here.

**Review by dafedafe — APPROVED** [review]
Thanks @jaskarth!

**Comment by eme64** [comment]
How will this mesh with the changes that @merykitty is doing with the KnownBits, in the work of https://github.com/openjdk/jdk/pull/17508 etc?

**Comment by eme64** [comment]
I think there are still quite a few unaddressed review comments above, so I'll hold off with reviewing myself. Feel free to ping me if you need another review.

**Comment by jaskarth** [comment]
@eme64 I was thinking that this PR and the associated IR tests could serve as a baseline for future future work in this area, such as #17508, since it just uses the signed bounds we already have to do its analysis. My goal was to make the changes as simple as possible, so that we can see some benefits short term while we work on the long-term solution for better type inference. Originally I just had the `[0..C]` case since I saw that it showed up a lot in real-world code, but I kept finding generalizations and eventually it snowballed :sweat_smile:

**Comment by chhagedorn** [comment]
I agree that we should keep this RFE simple. And we are just using thing that we already have. So, we could just go with the optimizations that you currently have (if you like to apply the few simple improvement suggestions, you can already do that) and follow up with future RFEs to cover more cases.

**Comment by jaskarth** [comment]
@chhagedorn I've pushed a new version that should address all the comments from code review. I ended up handling the case with two negative ranges as well because it didn't add too much code complexity, but I agree that we should just use  the optimizations we have and not add more to this PR.

**Review by chhagedorn — APPROVED** [review]
Thanks for the update, looks good to me! I'll give this another spinning in our testing over the weekend (will only be able to report back on Tuesday since Monday is a public holiday here).

**Review by merykitty — APPROVED** [review]

**Comment by chhagedorn** [comment]
Testing looked good!

**Comment by jaskarth** [comment]
Thanks a lot for the testing, and thank you everyone for the reviews!

/integrate

**Pull request closed without merging**
